design-proposals/instancetype-controllerrevision-migration.md

```diff
@@ -0,0 +1,228 @@
+
+[1] This initial step will be removed in the future once all instance type
+`ControllerRevisions` are labelled with at least
+`instancetype.kubevirt.io/object-version` making this a simple lookup using said
+label.
+
+## `ControllerRevisionUpgrade` controller
+
+Once stashed `instancetype.kubevirt.io` objects have been identified by the
+`ControllerRevisionUpgradeTrigger` controller and a
+`ControllerRevisionUpgrade` object submitted this controller then
+attempts to upgrade the object to the latest version.
+
+```go
+// ControllerRevisionUpgrade encapsulates a specific upgrade of a stashed ControllerRevision instance type object to the latest available version
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient
+type ControllerRevisionUpgrade struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControllerRevisionUpgradeSpec `json:"spec,omitempty"`
+	Status ControllerRevisionUpgradeStatus `json:"status,omitempty"`
+}
+
+type ControllerRevisionUpgradeSpec struct {
+	// Name of the ControllerRevision to migrate
+	TargetName string `json:"targetName"`
+}
+```
+
+* Watch for `ControllerRevisionUpgrade` requests
+
+* Extract and convert a stashed `instancetype.kubevirt.io` object within a given
+  `ControllerRevision` to the latest version
+
+* Recreate [2] the `ControllerRevision` and store the upgraded object
+
+* Patch the `Owner` object with a reference to the new `ControllerRevision`
+
+* Remove the original `ControllerRevision`
+
+[2] The object within a `ControllerRevisions` is immutable so we have to create
+a fresh `ControllerRevision` for the upgraded object.
+
+## Scalability
+
+
+## Update/Rollback Compatibility
+
+Once the initial `ControllerRevision` is deleted there will be no rollback
+possible to the original object.
+
+## Functional Testing Approach
+
+* Tests should populate an environment with `ControllerRevisoins` containing all
+  of the currently supported versions (`v1alpha{1,2}`, `v1beta1` etc..) and
+  migrate these to the latest version.
+
+## Implementation Phases
+
+* Introduce `ControllerRevisionUpgrade` CRD and controller
+
+* Introduce `ControllerRevisionUpgradeTrigger` CRD and controller
+
+## Alternatives
+
+### Upgrading on VirtualMachine resync
+
+This would rely on all `VirtualMachines` being resync'd periodically or on
+upgrade to first check and then upgrade any associated
+`instancetype.kubevirt.io` ControllerRevisions.
+
+*TODO* Finish the following section
+
+For
+
+* This would avoid the need to add additional CRDs and controllers to the codebase
+
+* `VirtualMachines` could be labelled to highlight that no upgrade was required *or* the upgrade was complete?
+
+Against
+
+* Difficult to tell when all `VirtualMachines` had been upgraded without additional labels
+
+* `VirtualMachineStatus` would need to be extended to capture failures
+
+* This would add overhead to each resync of a VirtualMachine using
```

Comment:
This could be handled by a new controller but without the need for CRD as well if we find this argument appealing.